File: 0Leetcode/0133-clone-graph/0133-clone-graph.cpp
class Solution {
public:
    Node* cloneGraph(Node* node) {
        unordered_map<Node*, Node*> oldtonew;
        return dfs(node, oldtonew);
    }

    Node* dfs(Node* node, unordered_map<Node*, Node*>& oldtonew) {
        if (!node) {
            return nullptr;
        }
        if (oldtonew.count(node)) {
            return oldtonew[node];
        }
        Node* copy = new Node(node->val);
        oldtonew[node] = copy;
        for (Node* n : node->neighbors) {
            copy->neighbors.push_back(dfs(n, oldtonew));
        }
        return copy;
        
    }
};

File: 0Leetcode/0133-clone-graph/0133-clone-graph.java
class Solution {
    Node []visited;
    public Node cloneGraph(Node node) {
        if(node==null) return null;
        visited = new Node[101];
        return dfs(node);
    }
    public Node dfs(Node node){
        if(visited[node.val] != null) return visited[node.val];
        visited[node.val] = new Node(node.val);
        for(Node n: node.neighbors) visited[node.val].neighbors.add(dfs(n));
        return visited[node.val];
    }
}